renice: make renice posix-compliant (fixes #1892)

kraptor opened the pull request:
As noted in bug #1892, renice is not 100% posix compliant as the '-n' option should update the priority 'relative' to the current priority.

This patch addresses this:
- donice() has now a new param to denote if priority is absolute or relative.
- main() interprets '-n' as relative, but leaves '--priority' as absolute.
- manpage is updated to display this changes.

After this patch:
- Calling renice with '-n 2' on a process with current priority '2', will try to update it to priority '4'.
- Calling renice with '--priority 2' on a process with current priority '2', leaves the process at priority '2'.

No other changes are made.

karelzak commented:
How many people already use -n to specify absolute value? 

We have supported this since the year 2009. Yes, it was added incorrectly as an absolute rather than a relative setting (my mistake ...), but it seems pretty late to fix it now.

What about adding getenv("POSIXLY_CORRECT") to implement -n in the backwardly incompatible relative way?

kraptor commented:
Can someone (@karelzak??) re-trigger the CI? Failures aren't related to the changes.

kraptor commented:
> How many people already use -n to specify absolute value?
> 
> We have supported this since the year 2009. Yes, it was added incorrectly as an absolute rather than a relative setting (my mistake ...), but it seems pretty late to fix it now.
> 
> What about adding getenv("POSIXLY_CORRECT") to implement -n in the backwardly incompatible relative way?

Yeah, we can do that, let me do it.

kraptor commented:
Ok, done. Added `--relative`, too. Updated all docs. Please review, I'm not a native English speaker and most probably the manpage/help messages can be worded better.

t-8ch commented on sys-utils/renice.c:
```diff
@@ -135,9 +145,28 @@ int main(int argc, char **argv)
+		{
+			// Fully conform to posix only if POSIXLY_CORRECT is
+			// set in the environment. If not, use the absolute
+			// value as it's been used (incorectly) since 2009
```
"incorrectly"

kraptor replied:
Good catch. I'll correct that. Thanks.

t-8ch commented on sys-utils/renice.1.adoc:
```diff
@@ -46,16 +46,22 @@ renice - alter priority of running processes
-*-n*, *--priority* _priority_::
-Specify the scheduling _priority_ to be used for the process, process group, or user. Use of the option *-n* or *--priority* is optional, but when used it must be the first argument.
+*-n* _priority_::
+Specify the *absolute* scheduling _priority_ to be used for the process, process group, or user. Use of the option *-n* is optional, but when used it must be the first argument. *Note* that the argument is *absolute* for historical reasons. To make this flag 100% posix compliant, define POSIXLY_CORRECT in the environement and it will use a *relative* value instead.
```
"environment"

t-8ch commented on sys-utils/renice.1.adoc:
```diff
@@ -46,16 +46,22 @@ renice - alter priority of running processes
+Specify the *absolute* scheduling _priority_ to be used for the process, process group, or user. Use of the option *-n* is optional, but when used it must be the first argument. *Note* that the argument is *absolute* for historical reasons. To make this flag 100% posix compliant, define POSIXLY_CORRECT in the environement and it will use a *relative* value instead.
+
+*--priority* _priority_::
+Specify the *absolute* scheduling _priority_ to be used for the process, process group, or user. Use of the option *--priority* is optional, but when used it must be the first argument.
```
The descriptions for `-n`, `--priority`, `--relative` are very repetitive.

kraptor replied:
I'll rework them, no problem.

t-8ch commented on sys-utils/renice.1.adoc:
```diff
@@ -46,16 +46,22 @@ renice - alter priority of running processes
 == SYNOPSIS
 
-*renice* [*-n*] _priority_ [*-g*|*-p*|*-u*] _identifier_...
+*renice* [*-n|--priority|--relative*] _priority_ [*-g*|*-p*|*-u*] _identifier_...
```
What's the point of `--priority`? It also seems a bit inconsistent wrt `--relative`.

kraptor replied:
Well, `--priority` was already there before and I left it as it was as I didn't want to break backwards compatibility. I just added `--relative`, should anyone want to use the "relative" version without setting up an environment var. Do you mind to explain a bit more what I need to do? Thanks.

t-8ch replied:
I didnt't realize tht `--priority` alrrady exists. Please disregard this comment.

kraptor replied:
Updated everything following your suggestions and squashed all commits.

kraptor commented:
I made all requested changes but the last, as I'm not sure what is requested. Thanks for the review. Let me know if there is anything else to do. Thanks!

kraptor commented:
Everything should be fine now. Please, review wording of manpage in last commit. Thanks!

kraptor commented:
Is there anything else to do on my side @karelzak ?

karelzak commented on sys-utils/renice.c:
```diff
@@ -89,13 +95,18 @@ static int getprio(const int which, const int who, int *prio)
+	if (relative)
+	    newprio += prio;
+
+	if (setpriority(which, who, newprio) < 0) {
```
This is not correct. The code does not modify priority if the `relative` is not set (as you call setpriority() with newprio=oldprio). I guess we need the following:
```
if (relative)
   prior = oldprio + prio;

if (setpriority(which, who, prio) < 0)
```

karelzak commented on sys-utils/renice.c:
```diff
@@ -59,16 +59,22 @@ static void __attribute__((__noreturn__)) usage(void)
-	fputs(_(" -n, --priority <num>   specify the nice value\n"), out);
+	fputs(_(" -n <num>               specify the nice value\n"), out);
+	fputs(_("                          If POSIXLY_CORRECT flag is set in environment\n"), out);
+	fputs(_("                          environment then the priority is 'relative'\n"), out);
```
It seems like a typo, `in environment environment then`

kraptor replied:
prio is const, so I updated the code to use newprio instead, but otherwise you were right and it is now corrected.

karelzak commented:
@kraptor, sorry for the delay, and thanks for your work. I was offline during Christmas time.

 The concept is fine, but please correct the above-mentioned mistakes.

kraptor commented:
> @kraptor, sorry for the delay, and thanks for your work. I was offline during Christmas time.
> The concept is fine, but please correct the above-mentioned mistakes.

What a silly mistake did I make... should be corrected now. :ashamed:
